examples/research_projects/jax-projects/README.md

```diff
@@ -383,13 +383,212 @@ official [flax example folder](https://github.com/huggingface/transformers/tree/
+      
+      params_rng, dropout_rng = jax.random.split(rng)
+      rngs = {"params": params_rng, "dropout": dropout_rng}
+      
+      params = self.module.init(rngs, inputs)["params"]
+      return params
+   
+   def __call__(self, inputs, params: dict = None):
+      params = {"params": params or self.params}
+      outputs = self.module.apply(params, jnp.array(inputs))
+      return outputs
+```
+
+
+Now we can define our model class as follows.
+
+```python
+class FlaxMLPModel(FlaxMLPPreTrainedModel):
+   module_class = FlaxMLPModule
+```
+
+Now the `FlaxMLPModel` will have a similar interface as that of other Transformers models.
+
+So the important point to remember is that the `model` is not an instance of `nn.Moudle`; it's an abstract class, like a container that holds a flax module, its parameters and provides convenient methods for initialization and forward pass. So our flax module is still stateless and immutable, but now we have a simple interface to work with. Feel free to take a look at the code to see how exactly this is implemented for ex. [`modeling_flax_bert.py`](https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/modeling_flax_bert.py)
+
+Another significant difference between Flax and PyTorch models is that `loss` is not computed inside the model, and none of the flax models accept `labels` argument.
+This is because in JAX to get the gradients that are required for the backward pass; we need to transform the loss function using the `grad` or `value_and_grad` transformations.
+So we decouple this from the modeling code and write the loss functions in training scripts.
+
 ### How to use flax models and example scripts
 
-TODO (should be filled by 29.06.)
+Let's first see how to load, save and do inference with flax models. As explained in the above section, all flax models in Transformers have similar API to PyTorch models, so we can use the familiar `from_pretrained` and `save_pretrained` methods to load and save flax models.
 
-### Flax design philosophy in transformers
+Let's use the base `FlaxRobertaModel` without any heads as an example.
+
+```python
+from transformers import FlaxRobertaModel, RobertaTokenizerFast
+import jax
+
+tokenizer = RobertaTokenizerFast.from_pretrained("roberta-base")
+inputs = tokenizer("JAX/Flax is amazing ", padding="max_length", max_length=128, return_tensors="np")
+
+model = FlaxRobertaModel.from_pretrained("julien-c/dummy-unknown")
+
+@jax.jit
+def run_model(input_ids, attention_mask):
+   # run a forward pass, should return an object `FlaxBaseModelOutputWithPooling`
+   return model(input_ids, attention_mask)
+
+outputs = run_model(**inputs)
+```
+
+We use `jit` to compile the function to get maximum performance. Note that in the above example, we set `padding=max_length` to pad all examples to the same length because JAX transformations like `jit` or `pmap` compile the function and expect static shapes. If the batches are not
```

Comment:
```suggestion
We use `jit` to compile the function to get maximum performance. Note that in the above example, we set `padding=max_length` to pad all examples to the same length. We do this because JAX's compiler has to recompile a function everytime its input shape changes - in a sense a compiled function is not only defined by its code but also by its input and output shape. It is usually much more effective to pad the input to be of a fixed static shape than having to recompile every the function multiple times.
```